junit5/src/main/java/cz/xtf/junit5/listeners/ProjectCreator.java

```diff
@@ -1,92 +1,76 @@
+import org.junit.platform.engine.FilterResult;
+import org.junit.platform.engine.TestDescriptor;
+import org.junit.platform.launcher.PostDiscoveryFilter;
 import org.junit.platform.launcher.TestExecutionListener;
 import org.junit.platform.launcher.TestPlan;
 
 import cz.xtf.core.config.OpenShiftConfig;
-import cz.xtf.core.openshift.OpenShift;
+import cz.xtf.core.config.XTFConfig;
+import cz.xtf.core.context.TestCaseContext;
+import cz.xtf.core.namespace.NamespaceManager;
 import cz.xtf.core.openshift.OpenShifts;
-import cz.xtf.core.waiting.SimpleWaiter;
 import cz.xtf.junit5.config.JUnitConfig;
-import io.fabric8.kubernetes.api.builder.Visitor;
-import io.fabric8.kubernetes.api.model.NamespaceBuilder;
-import io.fabric8.kubernetes.client.KubernetesClientException;
 import lombok.extern.slf4j.Slf4j;
 
 @Slf4j
-public class ProjectCreator implements TestExecutionListener {
-    private static final OpenShift openShift = OpenShifts.master();
+public class ProjectCreator
+        implements TestExecutionListener, BeforeAllCallback, AfterAllCallback, PostDiscoveryFilter {
 
     @Override
     public void testPlanExecutionStarted(TestPlan testPlan) {
-        createProject();
+        NamespaceManager.createIfDoesNotExistsProject(XTFConfig.get(OPENSHIFT_NAMESPACE));
     }
 
     @Override
-    public void testPlanExecutionFinished(TestPlan testPlan) {
-        if (JUnitConfig.cleanOpenShift()) {
-            deleteProject();
-        }
+    public void beforeAll(ExtensionContext context) {
+        // todo consider to move setting testcase context in separate extension
+        setTestExecutionContext(context);
+
+        log.debug("BeforeAll - Test case: " + context.getTestClass().get().getName() + " running in thread name: "
+                + Thread.currentThread().getName()
+                + " will use namespace: " + OpenShifts.master().getNamespace() + " - thread context is: "
+                + TestCaseContext.getRunningTestCaseName());
+        NamespaceManager.createIfDoesNotExistsProject();
+    }
+
+    private void setTestExecutionContext(ExtensionContext context) {
+        TestCaseContext.setRunningTestCase(context.getTestClass().get().getName());
     }
 
-    /**
-     * Delete the current project
-     * 
-     * @return boolean, true if the project has been deleted, false otherwise
-     */
-    public static boolean deleteProject() {
-        boolean deleted = openShift.deleteProject();
-        // For multi-module maven projects, other modules may attempt to crate project requests immediately after this modules deleteProject
-        if (deleted) {
-            BooleanSupplier bs = () -> openShift.getProject() == null;
-            new SimpleWaiter(bs, TimeUnit.MINUTES, 2, "Waiting for old project deletion").waitFor();
+    @Override
+    public void afterAll(ExtensionContext context) {
+        if (OpenShiftConfig.useNamespacePerTestcase() && JUnitConfig.cleanOpenShift()) {
```

Comment:
Couldn't be `OpenShiftConfig.useNamespacePerTestcase()` logic hidden in `NamespaceManager.deleteProject()` ? 

Seems to me yes. Just delete `NamespaceManager.getNamespace()` and `NamespaceManager.getNamespace()` will return correct namespace to delete.

Reply:
Right, fixed.